src/sinks/prometheus/exporter.rs

```diff
@@ -160,39 +179,160 @@ impl SinkConfig for PrometheusCompatConfig {
+            }
+            MetricValue::AggregatedSummary { quantiles, .. } => {
+                Some(quantiles.iter().map(|q| q.quantile).collect())
+            }
+            _ => None,
+        };
+
+        Self(
+            metric.series().clone(),
+            metric.kind(),
+            discriminant(metric.value()),
+            bounds,
+        )
+    }
+}
+
+impl PartialEq for MetricRef {
+    fn eq(&self, other: &Self) -> bool {
+        self.0 == other.0 && self.1 == other.1 && self.2 == other.2 && self.3 == other.3
+    }
+}
+
+impl Eq for MetricRef {}
+
+impl Hash for MetricRef {
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        self.0.hash(state);
+        self.1.hash(state);
+        self.2.hash(state);
+        if let Some(bounds) = &self.3 {
+            for bound in bounds {
+                bound.to_bits().hash(state);
+            }
+        }
+    }
+}
+
+struct PrometheusExporterMetricNormalizer {
+    distributions_as_summaries: bool,
+    buckets: Vec<f64>,
+}
+
+impl MetricNormalize for PrometheusExporterMetricNormalizer {
+    fn apply_state(&mut self, state: &mut MetricSet, metric: Metric) -> Option<Metric> {
+        let new_metric = match metric.value() {
+            MetricValue::Distribution { .. } => {
+                // Convert the distribution as-is, and then let the normalizer absolute-ify it.
+                let (series, data, metadata) = metric.into_parts();
+                let (ts, kind, value) = data.into_parts();
+
+                let new_value = match value {
+                    MetricValue::Distribution { samples, .. } => {
+                        if self.distributions_as_summaries {
+                            // We use a sketch when in summary mode because they're actually able to be
+                            // merged and provide correct output, unlike the aggregated summaries that
+                            // we handle from _sources_ like Prometheus.  The collector code already
+                            // renders sketches as an aggregated summary to begin with, so we're gucci.
+                            let mut sketch = AgentDDSketch::with_agent_defaults();
+                            for sample in samples {
+                                sketch.insert_n(sample.value, sample.rate);
+                            }
+
+                            MetricValue::Sketch {
+                                sketch: MetricSketch::AgentDDSketch(sketch),
+                            }
+                        } else {
+                            let (buckets, count, sum) =
+                                samples_to_buckets(samples.iter(), &self.buckets);
+
+                            MetricValue::AggregatedHistogram {
+                                buckets,
+                                count,
+                                sum,
+                            }
+                        }
+                    }
```

Comment:
This almost looks like it should be a method or pair of methods on `MetricValue`, ie `distribution_to_sketch` and `distribution_to_histogram` possibly with a `distribution_to_something(as_summary: bool)` wrapper.

Reply:
I definitely agree that they should be methods somewhere other than directly in the exporter, since it's fairly generic code.

I can try throwing them on `MetricValue` and seeing how that looks.

Reply:
Addressed in `91155f808` by moving these to `MetricValue` directly, as fallible conversions.